plugins/teststeps/teststeps_test.go

```diff
@@ -0,0 +1,369 @@
+// LICENSE file in the root directory of this source tree.
+
+package teststeps
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/facebookincubator/contest/pkg/cerrors"
+	"github.com/facebookincubator/contest/pkg/logging"
+	"github.com/facebookincubator/contest/pkg/target"
+	"github.com/facebookincubator/contest/pkg/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type data struct {
+	done          chan struct{}
+	cancel, pause chan struct{}
+	inCh, outCh   chan *target.Target
+	errCh         chan cerrors.TargetError
+	stepChans     test.TestStepChannels
+}
+
+func newData() data {
+	inCh := make(chan *target.Target)
+	outCh := make(chan *target.Target)
+	errCh := make(chan cerrors.TargetError)
+	return data{
+		done:   make(chan struct{}, 1),
+		cancel: make(chan struct{}),
+		pause:  make(chan struct{}),
+		inCh:   inCh,
+		outCh:  outCh,
+		errCh:  errCh,
+		stepChans: test.TestStepChannels{
+			In:  inCh,
+			Out: outCh,
+			Err: errCh,
+		},
+	}
+}
+
+func TestForEachTargetOneTarget(t *testing.T) {
+	d := newData()
+	fn := func(cancel, pause <-chan struct{}, tgt *target.Target) error {
+		log.Printf("Handling target %+v", tgt)
+		return nil
+	}
+	go func() {
+		d.inCh <- &target.Target{Name: "target001"}
+		// signal end of input
+		d.inCh <- nil
+	}()
+	go func() {
+		for {
+			select {
+			case <-d.done:
+				break
+			case tgt := <-d.outCh:
+				log.Printf("Step for target %+v completed as expected", tgt)
+			case err := <-d.errCh:
+				t.Errorf("Expected no error but got one: %v", err)
+			}
+		}
+	}()
+	err := ForEachTarget("test_one_target ", d.cancel, d.pause, d.stepChans, fn)
+	d.done <- struct{}{}
+	require.NoError(t, err)
+}
+
+func TestForEachTargetOneTargetAllFail(t *testing.T) {
+	d := newData()
+	fn := func(cancel, pause <-chan struct{}, t *target.Target) error {
+		log.Printf("Handling target %+v", t)
+		return fmt.Errorf("error with target %+v", t)
+	}
+	go func() {
+		d.inCh <- &target.Target{Name: "target001"}
+		// signal end of input
+		d.inCh <- nil
+	}()
+	go func() {
+		for {
+			select {
+			case <-d.done:
+				break
+			case tgt := <-d.outCh:
+				t.Errorf("Step for target %+v expected to fail but completed successfully instead", tgt)
+			case err := <-d.errCh:
+				log.Printf("Step for target failed as expected: %v", err)
+			}
+		}
+	}()
+	err := ForEachTarget("test_one_target ", d.cancel, d.pause, d.stepChans, fn)
+	d.done <- struct{}{}
+	require.NoError(t, err)
+}
+
+func TestForEachTargetTenTargets(t *testing.T) {
+	d := newData()
+	fn := func(cancel, pause <-chan struct{}, tgt *target.Target) error {
+		log.Printf("Handling target %+v", tgt)
+		return nil
+	}
+	go func() {
+		for i := 0; i < 10; i++ {
+			d.inCh <- &target.Target{Name: fmt.Sprintf("target%00d", i)}
+		}
+		// signal end of input
+		d.inCh <- nil
+	}()
+	go func() {
+		for {
+			select {
+			case <-d.done:
+				break
```

Comment:
It seems to be you meant `return` here (and in other tests).

Reply:
Most of the tests were created by @insomniacslk, but I'll check :)